### PR TITLE
feat: server-side SDK 

### PR DESCRIPTION
**Pull request by cwaldren-ld**
This PR exists to verify the linting, compilation, and test CI steps for the work-in-progress Server-Side SDK.

This feature branch does not yet represent the finished product but rather an alpha. Once merged, additional work is needed to onboard to release please.

### CODE REVIEW

File: .github/workflows/common.yml
Comment (cwaldren-ld) [inline .github/workflows/common.yml:9]: Can remove `server-side` once merged.

File: libs/internal/src/events/asio_event_processor.cpp
Comment (cwaldren-ld) [inline libs/internal/src/events/asio_event_processor.cpp:287]: TODO: Could probably achieve this with a constructor instead.

File: libs/client-sdk/README.md
Comment (cwaldren-ld) [inline libs/client-sdk/README.md:10]: TODO: The readme currently lists the cmake variables used to control the build. Should we move this into the top-level README, or copy into the server-sdk's README?


 Right now they are all general rather than sdk-specific.

File: libs/internal/src/serialization/json_flag.cpp
Comment (cwaldren-ld) [inline libs/internal/src/serialization/json_flag.cpp:230]: TODO

File: libs/internal/src/serialization/json_rule_clause.cpp
Comment (cwaldren-ld) [inline libs/internal/src/serialization/json_rule_clause.cpp:126]: TODOs

File: libs/server-sdk/src/all_flags_state/json_all_flags_state.cpp
Comment (cwaldren-ld) [inline libs/server-sdk/src/all_flags_state/json_all_flags_state.cpp:22]: TODO: switching these to WriteMinimal

File: libs/server-sdk/src/client_impl.cpp
Comment (cwaldren-ld) [inline libs/server-sdk/src/client_impl.cpp:54]: TODO

File: libs/internal/src/serialization/json_flag.cpp
Comment (cwaldren-ld) [reply libs/internal/src/serialization/json_flag.cpp:230]: Thinking about this, it feels like the long term solution shouldn't rely on "SDK deserializes data, then reserializes for persistence based on its in-memory model". Otherwise, old SDKs that aren't aware of new fields will drop them and "pollute" or otherwise downgrade the data. Instead, they should pass-through to the storage - save the CPU cycles and allow newer clients (including newer versions of the same SDK) to read the complete data.

Comment (kinyoklion) [reply libs/internal/src/serialization/json_flag.cpp:230]: I agree with this conceptually, but it has a couple problems which I tried to reason through when we started the V2 process. We need to update individual records and we likely do not want to store or re-create the textual record in order to update. (double memory, or base storage in a non-typed form, like keeping the boost::json representation).

We get an init, we save that to the persistent store. 

We get a patch for a flag in that init, what do we do? Read back the json, deserialize it, do patch comparisons, update it? Ok, now how do we update the `allFlags` key? We cannot really patch JSON without putting into some representable form, changing it, and then serializing it. Basically we have parallel implementations of the patch update process.

I think that we possibly could do something like this, but it would require changing how we store the representation. I am not sure how yet. Certainly dropping the idea of the possibility of a store without the ability to scan, which is where the possibility of an 'all' key comes from.

Comment (kinyoklion) [reply libs/internal/src/serialization/json_flag.cpp:230]: So, "all must be an aggregate of all the entries for the specified type". Which would hold true for dynamodb and redis.

File: libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp
Comment (cwaldren-ld) [inline libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp:15]: Calling out as a theoretical breaking change.

 In previous Client SDK, you could call `IDataSourceStatusProvider::Status()` and get back a `class client_side::data_sources::DataSourceStatus`.

Now, you get back a `class common::data_sources::DataSourceStatusBase<DataSourceState>`. 

These classes should have the same API, but the concrete type has changed. 

If the caller did:
```c++
auto status = sdk.DataSourceStatus().Status();
if (auto err = status.LastError()) {
   ...
}
```
then this should compile without any changes. But if they did:
```
client_side::data_sources::DataSourceStatus status = sdk.DataSourceStatus().Status();
```
that's going to require a change.

We have [some](https://github.com/launchdarkly/cpp-sdks/releases/tag/launchdarkly-cpp-client-v3.0.6) [precedent ](https://github.com/launchdarkly/cpp-sdks/releases/tag/launchdarkly-cpp-client-v3.0.5)for "non-major to fix a bug in the API definition".

**Comment by cwaldren-ld** [comment]
I've reviewed all files, fixed a bunch of outstanding TODOs, and got this branch all caught up with main.

I'm feeling ready to merge pending a backwards compatibility concern I've called out in comments.

File: libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp
Comment (cwaldren-ld) [reply libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp:15]: Actually, this might be fine. Since `client_side::data_sources::DataSourceStatus` is a type alias, and the new type didn't exist, it should still compile. It's definitely ABI breaking but we don't claim to support a stable C++ ABI.

Comment (cwaldren-ld) [reply libs/client-sdk/include/launchdarkly/client_side/data_source_status.hpp:15]: The following code compiles without issue on the existing client, and the client in this PR:
```c++
  client_side::data_sources::DataSourceStatus ds_status =
        client.DataSourceStatus().Status();

    std::optional<client_side::data_sources::DataSourceStatus::ErrorInfo>
        last_err = ds_status.LastError();

    client_side::data_sources::DataSourceStatus::DataSourceState state =
        ds_status.State();

    client_side::data_sources::DataSourceStatus::DateTime date =
        ds_status.StateSince();
```

File: libs/internal/src/serialization/json_flag.cpp
Comment (cwaldren-ld) [reply libs/internal/src/serialization/json_flag.cpp:228]: Added a todo link into the codebase for this. Deserves a more extended discussion, for sure.

**Review by kinyoklion — APPROVED** [review]

**Pull request merged**
